usersession/agent/rest_api.go

```diff
@@ -197,3 +211,105 @@ func postServiceControl(c *Command, r *http.Request) Response {
+	}
+
+	if mediaType != "application/json" {
+		return BadRequest("unknown content type: %s", contentType)
+	}
+
+	charset := strings.ToUpper(params["charset"])
+	if charset != "" && charset != "UTF-8" {
+		return BadRequest("unknown charset in content type: %s", contentType)
+	}
+
+	decoder := json.NewDecoder(r.Body)
+
+	// pendingSnapRefreshInfo holds information about pending snap refresh provided by snapd.
+	type pendingSnapRefreshInfo struct {
+		InstanceName        string        `json:"instance-name"`
+		TimeRemaining       time.Duration `json:"time-remaining,omitempty"`
+		BusyAppName         string        `json:"busy-app-name,omitempty"`
+		BusyAppDesktopEntry string        `json:"busy-app-desktop-entry,omitempty"`
+	}
+	var refreshInfo pendingSnapRefreshInfo
+	if err := decoder.Decode(&refreshInfo); err != nil {
+		return BadRequest("cannot decode request body into pending snap refresh info: %v", err)
+	}
+
+	conn, err := dbusutil.SessionBus()
+	if err != nil {
+		return SyncResponse(&resp{
+			Type:   ResponseTypeError,
+			Status: 500,
+			Result: &errorResult{
+				Message: fmt.Sprintf("cannot connect to the session bus: %v", err),
+			},
+		})
+	}
+	defer conn.Close()
+
+	// TODO: support desktop-specific notification APIs if they provide a better
+	// experience. For example, the GNOME notification API.
+	notifySrv := notification.New(conn)
+
+	// TODO: this message needs to be crafted better as it's the only thing guaranteed to be delivered.
+	summary := fmt.Sprintf(i18n.G("Pending update of %q snap"), refreshInfo.InstanceName)
+	var urgencyLevel notification.Urgency
+	var body, icon string
+	var hints []notification.Hint
+	if daysLeft := int(refreshInfo.TimeRemaining.Truncate(time.Hour).Hours() / 24); daysLeft > 0 {
+		urgencyLevel = notification.LowUrgency
+		body = fmt.Sprintf(i18n.NG(
+			"Close the app to avoid disruptions (%d day left)",
+			"Close the app to avoid disruptions (%d days left)", daysLeft), daysLeft)
+	} else if hoursLeft := int(refreshInfo.TimeRemaining.Truncate(time.Minute).Minutes() / 60); hoursLeft > 0 {
+		urgencyLevel = notification.NormalUrgency
+		body = fmt.Sprintf(i18n.NG(
+			"Close the app to avoid disruptions (%d hour left)",
+			"Close the app to avoid disruptions (%d hours left)", hoursLeft), hoursLeft)
+	} else if minutesLeft := int(refreshInfo.TimeRemaining.Truncate(time.Minute).Minutes()); minutesLeft > 0 {
+		urgencyLevel = notification.CriticalUrgency
+		body = fmt.Sprintf(i18n.NG(
+			"Close the app to avoid disruptions (%d minute left)",
+			"Close the app to avoid disruptions (%d minutes left)", minutesLeft), minutesLeft)
+	} else {
+		summary = fmt.Sprintf(i18n.G("Snap %q is refreshing now!"), refreshInfo.InstanceName)
+		urgencyLevel = notification.CriticalUrgency
+	}
+	hints = append(hints, notification.WithUrgency(urgencyLevel))
+	if refreshInfo.BusyAppDesktopEntry != "" {
+		hints = append(hints, notification.WithDesktopEntry(refreshInfo.BusyAppDesktopEntry))
```

Comment:
Please use `io.snapcraft.SessionAgent` as the desktop entry for the notification (the desktop entry and bus name added by PR #9497).  We don't want to be forging notifications from other applications.

Using the icon hint to show the snap's icon is fine though.

Reply:
DOne